src/libDirectoryService/FinalBlockPostProcessing.cpp

```diff
@@ -558,7 +561,16 @@ bool DirectoryService::ProcessFinalBlockConsensusCore(
   if (state == ConsensusCommon::State::DONE) {
     cv_viewChangeFinalBlock.notify_all();
     m_viewChangeCounter = 0;
-    ProcessFinalBlockConsensusWhenDone();
+    // Update the final block with the co-signatures from the consensus
+    m_finalBlock->SetCoSignatures(*m_consensusObject);
+
+    // Update the DS microblock with the same co-signatures from the consensus
+    // If we don't do this, DataSender won't be able to process it
+    m_mediator.m_node->m_microblock->SetCoSignatures(*m_consensusObject);
+
+    auto func = [this]() -> void { ProcessFinalBlockConsensusWhenDone(); };
```

Comment:
Why do we need to spawn another thread?

Reply:
Because the mutex `m_mutexConsensus` will be occupied until the consensus end if we didn't throw out a thread. While missing microblock fetching still requires the consensus to be in activated state.

Reply:
But `ProcessFinalBlockConsensusWhenDone` won't be called if `state == ConsensusCommon::State::ERROR` and while MB is being fetched.
So `m_mutexConsensus` is not needed right? or am i missing something

Reply:
still didn't get clear reason for spawning new thread
@ansnunez did you?